gemini/tuning/gemini_supervised_finetuning_text_classification.ipynb

```diff
@@ -0,0 +1,2295 @@
+      "outputs": [],
+      "source": [
+        "PROJECT_ID = \"[your-project-id]\"  # @param {type:\"string\"}\n",
+        "LOCATION = \"us-central1\"  # @param {type:\"string\"}\n",
+        "\n",
+        "vertexai.init(project=PROJECT_ID, location=LOCATION)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "C5r1Bhf7fcuA"
+      },
+      "outputs": [],
+      "source": [
+        "BUCKET_NAME = \"[yout-bucket-name]]\"  # @param {type:\"string\"}\n",
+        "BUCKET_URI = f\"gs://{BUCKET_NAME}\"\n"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "eSZS8QfOfcuA"
+      },
+      "source": [
+        "**warning:** Only if your bucket doesn't already exist: Run the following cell to create your Cloud Storage bucket.\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "QXxi2tC2fcuA"
+      },
+      "outputs": [],
+      "source": [
+        "! gsutil mb -l $LOCATION -p $PROJECT_ID $BUCKET_URI"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "pSZ8-hwafcuA"
+      },
+      "source": [
+        "Finally, validate access to your Cloud Storage bucket by examining its contents"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "YqCIzmWSfcuA"
+      },
+      "outputs": [],
+      "source": [
+        "! gsutil ls -al $BUCKET_URI"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "XAEAlYH2fcuA"
+      },
+      "source": [
+        "### Helper Functions"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "f6sph8etfcuA"
+      },
+      "source": [
+        "#### Batch Prediction - Helper functions"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "Fz4tGANnfcuA"
+      },
+      "outputs": [],
+      "source": [
+        "def backoff_hdlr(details):\n",
+        "    print(\"Backing off {wait:0.1f} seconds after {tries} tries \".format(**details))\n",
+        "\n",
+        "\n",
+        "def log_error(msg, *args):\n",
+        "    mp.get_logger().error(msg, *args)\n",
+        "    raise Exception(msg)\n",
+        "\n",
+        "\n",
+        "def handle_exception_threading(f):\n",
+        "    def applicator(*args, **kwargs):\n",
+        "        try:\n",
+        "            return f(*args, **kwargs)\n",
+        "        except:\n",
+        "            log_error(traceback.format_exc())\n",
+        "\n",
+        "    return applicator\n",
+        "\n",
+        "\n",
+        "@handle_exception_threading\n",
+        "@backoff.on_exception(backoff.expo, ResourceExhausted, max_tries=30, on_backoff=backoff_hdlr)\n",
+        "def _predict_message(message: str, model: GenerativeModel):\n",
```

Comment:
Add return type. Do the same for all functions and be sure to add parameter typing as well.